### PR TITLE
RSPEC-5665 Escape sequences should not be used in text blocks

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**

### CODE REVIEW

**Comment by margarita-nedzelska-sonarsource** [comment]
I've decided to report issues at line having improper character instead of the full Text Block, because it might have a lot of lines and it will be hard to locate the issue properly.

**Comment by margarita-nedzelska-sonarsource** [comment]
While working on this issue faced a problem with line numbers and reported a bug in JDT: https://bugs.eclipse.org/bugs/show_bug.cgi?id=570472

File: java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.sonar.java.collections.SetUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S5665")
+public class UnnecessaryEscapeSequencesInTextBlockCheck extends IssuableSubscriptionVisitor {
+  private static final String MESSAGE = "Remove this unnecessary escape sequence.";
+  private static final String TRIPLE_QUOTE_MESSAGE = "Use '\\\"\"\"' to escape \"\"\".";
+  private static final Set<String> ESCAPED = SetUtils.immutableSetOf("\\n", "\\t", "\\'", "\\\"");
```
Comment (sebastian-hungerecker-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java:35]: I wouldn't include `\t` in the list of escapes we complain about. I feel that using `\t` is a much better idea than putting literal tab characters into your string - whether it's inside a normal string literal or a text block.

File: java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.TEXT_BLOCK);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    LiteralTree textBlock = (LiteralTree) tree;
+    int startLine = textBlock.token().line();
+    String value = textBlock.value();
+    String[] lines = value.split("(\r?\n)|\r");
```
Comment (sebastian-hungerecker-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java:47]: The parentheses around `\r?\n` are redundant.

**Review by sebastian-hungerecker-sonarsource — APPROVED** [review]
See comments. LGTM otherwise.

File: java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java (hunk 1 above)
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/UnnecessaryEscapeSequencesInTextBlockCheck.java:35]: This does make sense, thanks

**Pull request merged**